### PR TITLE
making 10 language-independent ASR and CHiME-5 recipes

### PR DESCRIPTION
**Pull request by sw005320**
- Finished the 10 language-independent ASR task
- Finished the CHiME-5 task (the performance is really bad....)
- Added Nbest

### CODE REVIEW

**Comment by sw005320** [comment]
@kan-bayashi, can you find why the travis check is failed only when python 3.6?

File: src/nets/e2e_asr_attctc.py [hunk 1]
```diff
@@ -161,9 +161,16 @@ def __call__(self, data):
         '''
         # utt list of frame x dim
         xs = [i[1]['feat'] for i in data]
+        # remove 0-output-length utterances
+        tids = [d[1]['tokenid'].split() for d in data]
+        filtered_index = filter(lambda i: len(tids[i]) > 0, range(len(xs)))
+        if len(filtered_index) != len(xs):
```
Comment (kan-bayashi) [inline src/nets/e2e_asr_attctc.py]: You should add following line.
```
sorted_index = sorted(filtered_index, key=lambda i: -len(xs[i]))
```

**Comment by sw005320** [comment]
@kan-bayashi, @ShigekiKarita, can you quickly review them?
I have to release the CHiME-5 recipe on Monday.

File: README.md [hunk 2]
```diff
@@ -13,7 +13,7 @@ and also follows [Kaldi](http://kaldi-asr.org/) style data processing, feature e
   - Fast/accurate training with CTC/attention multitask training
   - CTC/attention joint decoding to boost monotonic alignment decoding
 - Encoder: VGG-like CNN + BLSTM or pyramid BLSTM
-- Attention: Dot product or location-aware attention
+- Attention: Dot product, location-aware attention, variants of multhead (pytorch only)
```
Comment (ShigekiKarita) [inline README.md]: typo: multhead -> multihead

File: egs/li10/asr1/cmd.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+# machine).  queue.pl works with GridEngine (qsub).  slurm.pl works
+# with slurm.  Different queues are configured differently, with different
+# queue names and different ways of specifying things like memory;
+# to account for these differences you can create and edit the file
+# conf/queue.conf to match your queue's configuration.  Search for
+# conf/queue.conf in http://kaldi-asr.org/doc/queue.html for more information,
+# or search for the string 'default_config' in utils/queue.pl or utils/slurm.pl.
+
+export train_cmd="queue.pl --mem 2G"
+export cuda_cmd="queue.pl --mem 8G --gpu 1 --config conf/gpu.conf"
```
Comment (ShigekiKarita) [inline egs/li10/asr1/cmd.sh]: this is the JHU conf isn't it?  it is nice to be the same to the other egs https://github.com/sw005320/espnet-1/blob/f553100c2c96c8715062f5bc5b168dffb3dc5a4f/egs/wsj/asr1/cmd.sh

File: src/asr/asr_chainer.py [hunk 4]
```diff
@@ -341,7 +341,12 @@ def recog(args):
     new_json = {}
     for name, feat in reader:
         logging.info('decoding ' + name)
-        y_hat = e2e.recognize(feat, args, train_args.char_list, rnnlm)
+        if args.beam_size == 1:
+            y_hat = e2e.recognize(feat, args, train_args.char_list, rnnlm)
+        else:
+            nbest_hyps = e2e.recognize(feat, args, train_args.char_list, rnnlm)
+            # get 1best and revmoe sos
```
Comment (ShigekiKarita) [inline src/asr/asr_chainer.py]: typo: revmoe -> remove

File: egs/chime5/asr1/cmd.sh [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+# machine).  queue.pl works with GridEngine (qsub).  slurm.pl works
+# with slurm.  Different queues are configured differently, with different
+# queue names and different ways of specifying things like memory;
+# to account for these differences you can create and edit the file
+# conf/queue.conf to match your queue's configuration.  Search for
+# conf/queue.conf in http://kaldi-asr.org/doc/queue.html for more information,
+# or search for the string 'default_config' in utils/queue.pl or utils/slurm.pl.
+
+export train_cmd="queue.pl --mem 2G"
+export cuda_cmd="queue.pl --mem 8G --gpu 1 --config conf/gpu.conf"
```
Comment (ShigekiKarita) [inline egs/chime5/asr1/cmd.sh]: ditto

**Comment by ShigekiKarita** [comment]
do you plan to add RESULTS of chime5? it is useful for sanity check of user's environment.

File: README.md (hunk 2 above)
Comment (sw005320) [reply README.md]: done

File: egs/li10/asr1/cmd.sh (hunk 3 above)
Comment (sw005320) [reply egs/li10/asr1/cmd.sh]: done

File: src/asr/asr_chainer.py (hunk 4 above)
Comment (sw005320) [reply src/asr/asr_chainer.py]: done

File: egs/chime5/asr1/cmd.sh (hunk 5 above)
Comment (sw005320) [reply egs/chime5/asr1/cmd.sh]: done

**Comment by sw005320** [comment]
I'm still computing the final result. It will be finished tomorrow (I hope)

**Comment by sw005320** [comment]
@ShigekiKarita Is it OK to merge the current version, and later I'll add `RESULTS`? I just want to make me feel safe by that.

File: egs/chime5/asr1/run.sh [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+    if ${do_delta}; then
+        expdir=${expdir}_delta
+    fi
+else
+    expdir=exp/${train_set}_${tag}
+fi
+mkdir -p ${expdir}
+
+# switch backend
+if [[ ${backend} == chainer ]]; then
```
Comment (kan-bayashi) [inline egs/chime5/asr1/run.sh]: No more switching.

File: egs/li10/asr1/run.sh [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+    if ${do_delta}; then
+        expdir=${expdir}_delta
+    fi
+else
+    expdir=exp/${train_set}_${tag}
+fi
+mkdir -p ${expdir}
+
+# switch backend
+if [[ ${backend} == chainer ]]; then
```
Comment (kan-bayashi) [inline egs/li10/asr1/run.sh]: No more switching.

File: egs/li10/asr1/run.sh [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+else
+    train_script=asr_train_th.py
+    decode_script=asr_recog_th.py
+fi
+
+if [ ${stage} -le 3 ]; then
+    echo "stage 3: Network Training"
+    ${cuda_cmd} ${expdir}/train.log \
+        ${train_script} \
+        --gpu ${gpu} \
```
Comment (kan-bayashi) [inline egs/li10/asr1/run.sh]: `${train_script}` -> `asr_train.py`

Comment (kan-bayashi) [reply egs/li10/asr1/run.sh]: add `--backend ${backend} \`

File: egs/chime5/asr1/run.sh [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+    train_script=asr_train_th.py
+    decode_script=asr_recog_th.py
+fi
+
+if [ ${stage} -le 4 ]; then
+    echo "stage 4: Network Training"
+
+    ${cuda_cmd} ${expdir}/train.log \
+        ${train_script} \
+        --gpu ${gpu} \
```
Comment (kan-bayashi) [inline egs/chime5/asr1/run.sh]: `${train_script}` -> `asr_train.py`
add `--backend ${backend}`

File: egs/chime5/asr1/run.sh [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+
+        # make json labels for recognition
+        data2json.sh --nlsyms ${nlsyms} ${data} ${dict} > ${data}/data.json
+
+        #### use CPU for decoding
+        gpu=-1
+
+        ${decode_cmd} JOB=1:${nj} ${expdir}/${decode_dir}/log/decode.JOB.log \
+            ${decode_script} \
+            --gpu ${gpu} \
```
Comment (kan-bayashi) [inline egs/chime5/asr1/run.sh]: `${decode_script}` -> `asr_recog.py`
add `--backend ${backend}`

File: egs/li10/asr1/run.sh [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        # make json labels for recognition
+        data2json.sh --nlsyms ${nlsyms} ${data} ${dict} > ${data}/data.json
+
+        #### use CPU for decoding
+        gpu=-1
+
+        ${decode_cmd} JOB=1:${nj} ${expdir}/${decode_dir}/log/decode.JOB.log \
+            ${decode_script} \
+            --gpu ${gpu} \
+            --recog-feat "$feats" \
```
Comment (kan-bayashi) [inline egs/li10/asr1/run.sh]: ${decode_script} -> asr_recog.py
add --backend ${backend}

File: README.md [hunk 12]
```diff
@@ -117,6 +117,17 @@ We list the character error rate (CER) and word error rate (WER) of major ASR ta
...
 
 
+## Chainer and Pytorch backends
+
+|           | Chainer | Pytorch |
+|-----------|:----:|:----:|
+| Performance | ◎ | ○ |
+| Speed | ○ | ◎ |
+| Multiple GPUs | no support | supported |
+| VGG-like encoder | supported | no support |
```
Comment (kan-bayashi) [inline README.md]: Is multiple gpu in pytorch already available?

Comment (kan-bayashi) [reply README.md]: If so, I can merge #109

**Comment by ShigekiKarita** [comment]
@sw005320 LGTM

**Comment by kan-bayashi** [comment]
LGTM

**Comment by sw005320** [comment]
@ShigekiKarita Can you test it in your environment?
@kan-bayashi Do you get CHiME-5?
You'll get it from http://spandh.dcs.shef.ac.uk/chime_challenge/download.html
Also I want you to test it. The current performance seems to be very bad, and need some improvements.

**Comment by ShigekiKarita** [comment]
OK. I will run this today later.

**Comment by kan-bayashi** [comment]
Sure!

**Pull request merged**